### ISSUE TITLE
Query pagination

### ISSUE THREAD

**Issue by lukehutch** [serverpod/serverpod#1057]
**Is your feature request related to a problem? Please describe.**

Serverpod needs to support some sort of query pagination.

Here are some alternatives for Postgres:

https://www.citusdata.com/blog/2016/03/30/five-ways-to-paginate/

The simplest (from the point of view of the API user) and most accurate method is to use cursors. However this makes the server stateful, and ties up server resources.

If a query is ordered, pagination is also fairly straightforward to implement without cursors, but this does not take into account database mutation.

**Describe the solution you'd like**

Ideally cursors would be available, expiring if unused for a certain period of time.

**Describe alternatives you've considered**

Right now only complete queries can be generated. Or, probably, the user could implement some custom pagination scheme. But Serverpod should provide some options so that users don't have to reinvent any wheels.

**Comment by Isakdl** [comment]
Agreed, quite common use-case as well. Would be cool to support several different ways as they have some tradeoffs.

**Comment by SandPod** [comment]
We have added a section to the documentation that details the current support for query pagination: https://docs.serverpod.dev/next/concepts/database/pagination

The guide includes a cursor based query pagination, utilizing the ids we always generate, and an offset query pagination (with the drawbacks mentioned that it doesn't take into account database mutations).

Both can be used without adding state to the server.

**Comment by lukehutch** [comment]
@SandPod Um, that is not a cursor! :-D

This is a cursor:

https://www.postgresql.org/docs/current/plpgsql-cursors.html

Your proposed "cursor" system breaks horribly if you include an `ORDER BY` clause in the query (or, in theory, if Postgres defaults to returning records in any order other than increasing by id).

**Comment by SandPod** [comment]
@lukehutch Very true! Those are not compatible with ordering.

Then let's keep this issue open as a feature request for "true" cursors in Serverpod.

Thanks 🙏

**Comment by lukehutch** [comment]
I also want to point out the performance implication of query pagination using the two methods used in the docs (by the way you need to change the word "cursor" to "alternative way to paginate" in the docs): these methods take O(N^2) time in the number of results fetched, because every time the next page is fetched, all previous pages have to be re-fetched first (the whole query is run from scratch up until the current page).

A true cursor is more costly for the database to maintain, but it scales linearly in the length of the result set.

**Comment by SandPod** [comment]
@lukehutch you are correct that that is true for the offset based pagination: https://planetscale.com/blog/mysql-pagination#performance-drawbacks-of-offset-limit-pagination.
(That information should also be added to the docs)

But as far as I understand the ID based pagination shouldn't suffer from the same drawback as a where query would be utilized there: https://planetscale.com/blog/mysql-pagination#cursor-pagination

**Comment by lukehutch** [comment]
@SandPod the docs just say it may not be performant, you need to say the performance will scale quadratically in the number of pages fetched, for offset-based pagination...

Index-based pagination should avoid the quadratic scaling, by taking a log(N) hit for an index tree lookup, and yes, it happens in a `WHERE` clause, so it can actually be handled pretty efficiently. It's just very inflexible when it comes to complex queries. I have some queries now that generate 3 pages of SQL, including multiple alternative ways of ordering the results depending on the search criteria -- and I can't use something as simple as an id-based inequality to paginate these results.

And again, you really (really) can't use the word "cursor" for index-based pagination, because "cursor" has a very specific meaning in the context of databases. Please at least change that part of the docs.
